src/ompl/geometric/planners/quotientspace/src/MultiQuotient.ipp

```diff
@@ -0,0 +1,338 @@
+{
+    std::vector<int> feasibleNodesPerLevel;
+    for (unsigned int k = 0; k < quotientSpaces_.size(); k++)
+    {
+        unsigned int Nk = quotientSpaces_.at(k)->getTotalNumberOfFeasibleSamples();
+        feasibleNodesPerLevel.push_back(Nk);
+    }
+    return feasibleNodesPerLevel;
+}
+
+template <class T>
+std::vector<int> MultiQuotient<T>::getDimensionsPerLevel() const
+{
+    std::vector<int> dimensionsPerLevel;
+    for (unsigned int k = 0; k < quotientSpaces_.size(); k++)
+    {
+        unsigned int Nk = quotientSpaces_.at(k)->getDimension();
+        dimensionsPerLevel.push_back(Nk);
+    }
+    return dimensionsPerLevel;
+}
+
+template <class T>
+MultiQuotient<T>::~MultiQuotient()
+{
+}
+
+template <class T>
+void MultiQuotient<T>::setup()
+{
+    Planner::setup();
+    for (unsigned int k = 0; k < stopAtLevel_; k++)
+    {
+        quotientSpaces_.at(k)->setup();
+    }
+    currentQuotientLevel_ = 0;
+}
+
+template <class T>
+void MultiQuotient<T>::setStopLevel(unsigned int level_)
+{
+    if (level_ > quotientSpaces_.size())
+    {
+        stopAtLevel_ = quotientSpaces_.size();
+    }
+    else
+    {
+        stopAtLevel_ = level_;
+    }
+    std::cout << "new stop level: " << stopAtLevel_ << " from " << quotientSpaces_.size() << std::endl;
+}
+
+template <class T>
+void MultiQuotient<T>::clear()
+{
+    Planner::clear();
+
+    for (unsigned int k = 0; k < quotientSpaces_.size(); k++)
+    {
+        quotientSpaces_.at(k)->clear();
+    }
+    currentQuotientLevel_ = 0;
+
+    while (!priorityQueue_.empty())
+        priorityQueue_.pop();
+    foundKLevelSolution_ = false;
+
+    solutions_.clear();
+    pdef_->clearSolutionPaths();
+    for (unsigned int k = 0; k < pdefVec_.size(); k++)
+    {
+        pdefVec_.at(k)->clearSolutionPaths();
+    }
+}
+
+template <class T>
+ob::PlannerStatus MultiQuotient<T>::solve(const ob::PlannerTerminationCondition &ptc)
+{
+    ompl::time::point t_start = ompl::time::now();
+
+    for (unsigned int k = currentQuotientLevel_; k < stopAtLevel_; k++)
+    {
+        foundKLevelSolution_ = false;
+
+        if (priorityQueue_.size() <= currentQuotientLevel_)
+            priorityQueue_.push(quotientSpaces_.at(k));
+
+        ob::PlannerTerminationCondition ptcOrSolutionFound([this, &ptc] { return ptc || foundKLevelSolution_; });
+
+        while (!ptcOrSolutionFound())
+        {
+            og::QuotientSpace *jQuotient = priorityQueue_.top();
+            priorityQueue_.pop();
+            jQuotient->grow();
+
+            bool hasSolution = quotientSpaces_.at(k)->hasSolution();
+            if (hasSolution)
+            {
+                ob::PathPtr sol_k;
+                quotientSpaces_.at(k)->getSolution(sol_k);
+                solutions_.push_back(sol_k);
+                if (DEBUG)
+                {
+                    double t_k_end = ompl::time::seconds(ompl::time::now() - t_start);
+                    std::cout << std::string(80, '#') << std::endl;
```

Comment:
Please use OMPL_DEBUG and similar logging macros defined in ompl/util/Console.h instead of printing directly on std::cout. (Not just here but everywhere else, too.)